File: server/controller/emby.go
package controller

import (
	"strconv"

	"github.com/MccRay-s/alist2strm/model/common/response"
	"github.com/MccRay-s/alist2strm/service"
	"github.com/gin-gonic/gin"
)

// EmbyController Emby 控制器
type EmbyController struct{}

// Emby 控制器实例
var Emby = &EmbyController{}

// GetLibraries 获取Emby媒体库列表
// @Summary 获取Emby媒体库列表
// @Description 获取Emby服务器中所有可用的媒体库
// @Tags Emby
// @Accept json
// @Produce json
// @Success 200 {object} response.Response{data=[]service.EmbyLibrary}
// @Failure 400 {object} response.Response
// @Router /api/emby/libraries [get]
func (ctrl *EmbyController) GetLibraries(c *gin.Context) {
	libraries, err := service.Emby.GetLibraries()
	if err != nil {
		response.FailWithMessage("获取Emby媒体库列表失败: "+err.Error(), c)
		return
	}
	response.SuccessWithData(libraries, c)
}

// RefreshLibrary 刷新指定Emby媒体库
// @Summary 刷新指定Emby媒体库
// @Description 触发刷新Emby中的特定媒体库
// @Tags Emby
// @Accept json
// @Produce json
// @Param id path string true "媒体库ID"
// @Success 200 {object} response.Response
// @Failure 400 {object} response.Response
// @Router /api/emby/libraries/{id}/refresh [post]
func (ctrl *EmbyController) RefreshLibrary(c *gin.Context) {
	// 从路径参数获取媒体库ID
	libraryID := c.Param("id")
	if libraryID == "" {
		response.FailWithMessage("媒体库ID不能为空", c)
		return
	}

	err := service.Emby.RefreshLibrary(libraryID)
	if err != nil {
		response.FailWithMessage("刷新媒体库失败: "+err.Error(), c)
		return
	}
	response.SuccessWithMessage("已成功触发媒体库刷新", c)
}

// RefreshAllLibraries 刷新所有Emby媒体库
// @Summary 刷新所有Emby媒体库
// @Description 触发刷新Emby中的所有媒体库
// @Tags Emby
// @Accept json
// @Produce json
// @Success 200 {object} response.Response
// @Failure 400 {object} response.Response
// @Router /api/emby/libraries/refresh [post]
func (ctrl *EmbyController) RefreshAllLibraries(c *gin.Context) {
	err := service.Emby.RefreshAllLibraries()
	if err != nil {
		response.FailWithMessage("刷新所有媒体库失败: "+err.Error(), c)
		return
	}
	response.SuccessWithMessage("已成功触发所有媒体库刷新", c)
}

// GetLatestMedia 获取Emby最新入库媒体
// @Summary 获取Emby最新入库媒体
// @Description 获取Emby服务器中最新添加的媒体文件
// @Tags Emby
// @Accept json
// @Produce json
// @Param limit query int false "返回结果数量限制，默认10条"
// @Success 200 {object} response.Response{data=[]service.EmbyLatestMedia}
// @Failure 400 {object} response.Response
// @Router /api/emby/latest [get]
func (ctrl *EmbyController) GetLatestMedia(c *gin.Context) {
	limitStr := c.DefaultQuery("limit", "10")
	limit, err := strconv.Atoi(limitStr)
	if err != nil || limit <= 0 {
		limit = 10
	}

	media, err := service.Emby.GetLatestMedia(limit)
	if err != nil {
		response.FailWithMessage("获取最新入库媒体失败: "+err.Error(), c)
		return
	}
	response.SuccessWithData(media, c)
}

// TestConnection 测试Emby服务器连接
// @Summary 测试Emby服务器连接
// @Description 测试Emby服务器的可用性和连接状态
// @Tags Emby
// @Accept json
// @Produce json
// @Success 200 {object} response.Response{data=response.EmbyConnectionTestResult}
// @Failure 400 {object} response.Response
// @Router /api/emby/test [get]
func (ctrl *EmbyController) TestConnection(c *gin.Context) {
	result, err := service.Emby.TestConnection()
	if err != nil {
		response.FailWithMessage("测试连接时发生错误: "+err.Error(), c)
		return
	}
	response.SuccessWithData(result, c)
}

// GetImage 获取Emby图片
// @Summary 获取Emby图片
// @Description 代理获取Emby服务器上的图片资源
// @Tags Emby
// @Accept json
// @Produce image/*
// @Param item_id path string true "项目ID"
// @Param image_type path string true "图片类型,例如:Primary,Backdrop等"
// @Param tag query string false "图片标签"
// @Param max_width query int false "最大宽度"
// @Param max_height query int false "最大高度"
// @Param quality query int false "图片质量"
// @Success 200 {file} binary "图片文件"
// @Failure 400 {object} response.Response
// @Router /api/emby/items/{item_id}/images/{image_type} [get]
func (ctrl *EmbyController) GetImage(c *gin.Context) {
	// 获取路径参数
	itemId := c.Param("item_id")
	imageType := c.Param("image_type")
	if itemId == "" || imageType == "" {
		response.FailWithMessage("项目ID和图片类型不能为空", c)
		return
	}

	// 获取查询参数
	tag := c.Query("tag")
	maxWidth, _ := strconv.Atoi(c.Query("max_width"))
	maxHeight, _ := strconv.Atoi(c.Query("max_height"))
	quality, _ := strconv.Atoi(c.Query("quality"))

	// 调用服务获取图片
	imageData, contentType, err := service.Emby.GetImage(itemId, imageType, tag, maxWidth, maxHeight, quality)
	if err != nil {
		response.FailWithMessage("获取图片失败: "+err.Error(), c)
		return
	}

	// 设置响应头
	c.Header("Content-Disposition", "inline")
	c.Data(200, contentType, imageData)
}

File: frontend/src/api/emby.ts
import { http } from './http'

// Emby 媒体库信息类型
interface EmbyLibrary {
  Id: string // 媒体库 ID
  Name: string // 媒体库名称
  Locations?: string[] // 媒体库位置路径列表
  CollectionType?: string // 集合类型(movies,tvshows,music等)
  ItemId?: string // 项目ID(兼容旧版)
  Guid?: string // 全局唯一标识符
  PrimaryImageItemId?: string // 主图像项目ID
  PrimaryImageTag?: string // 主图像标签
  RefreshProgress?: number // 刷新进度
  RefreshStatus?: string // 刷新状态
  ItemCount: number // 媒体库中的项目数量
  Type?: string // 媒体类型
  LastUpdate?: string // 最后更新时间
  LibraryOptions?: any // 库配置选项
}

// Emby 最新入库媒体信息类型
interface EmbyLatestMedia {
  Id: string
  Name: string
  Path: string
  Type: string
  DateCreated: string
  PremiereDate?: string
  ProductionYear?: number
  Overview?: string
}

// Emby 连接测试结果类型
interface EmbyConnectionTestResult {
  connected: boolean
  error?: string
  version?: string
  serverName?: string
  operatingSystem?: string
}

export class EmbyAPI {
  private baseUrl = '/emby'

  /**
   * 测试 Emby 服务器连接
   * @returns 连接测试结果
   */
  async testConnection() {
    return http.get<EmbyConnectionTestResult>(`${this.baseUrl}/test`)
  }

  /**
   * 获取 Emby 媒体库列表
   * @returns 媒体库列表
   */
  async getLibraries() {
    return http.get<EmbyLibrary[]>(`${this.baseUrl}/libraries`)
  }

  /**
   * 获取 Emby 最新入库媒体
   * @param limit 返回结果数量限制，默认 10
   * @returns 最新入库媒体列表
   */
  async getLatestMedia(limit: number = 10) {
    return http.get<EmbyLatestMedia[]>(`${this.baseUrl}/latest`, {
      params: { limit },
    })
  }

  /**
   * 刷新指定的 Emby 媒体库
   * @param libraryId 媒体库 ID
   * @returns 刷新结果
   */
  async refreshLibrary(libraryId: string) {
    return http.post(`${this.baseUrl}/libraries/${libraryId}/refresh`)
  }

  /**
   * 刷新所有 Emby 媒体库
   * @returns 刷新结果
   */
  async refreshAllLibraries() {
    return http.post(`${this.baseUrl}/libraries/refresh`)
  }

  /**
   * 获取单个 Emby 媒体库详细信息
   * @param libraryId 媒体库 ID
   * @returns 媒体库详细信息
   */
  async getLibraryDetails(libraryId: string) {
    return http.get<EmbyLibrary>(`${this.baseUrl}/libraries/${libraryId}`)
  }

  /**
   * 获取 Emby 图片的 URL
   * @param itemId 项目 ID
   * @param imageType 图片类型 (Primary, Backdrop, Logo等)
   * @param options 可选参数对象
   * @param options.tag 可选的图片标签
   * @param options.maxWidth 可选的最大宽度
   * @param options.maxHeight 可选的最大高度
   * @param options.quality 可选的图片质量
   * @returns 图片 URL
   */
  getImageUrl(
    itemId: string,
    imageType: string,
    options?: {
      tag?: string
      maxWidth?: number
      maxHeight?: number
      quality?: number
    },
  ): string {
    // 在开发环境使用完整服务器 URL，生产环境使用相对路径
    const prefix = import.meta.env.VITE_DIRECT_SERVER_URL === 'true'
      ? import.meta.env.VITE_API_BASE_URL || ''
      : ''

    let url = `${prefix}/api/emby/items/${itemId}/images/${imageType}`

    // 添加可选查询参数
    const params = new URLSearchParams()
    if (options?.tag) {
      params.append('tag', options.tag)
    }
    if (options?.maxWidth) {
      params.append('max_width', options.maxWidth.toString())
    }
    if (options?.maxHeight) {
      params.append('max_height', options.maxHeight.toString())
    }
    if (options?.quality) {
      params.append('quality', options.quality.toString())
    }

    const queryString = params.toString()
    if (queryString) {
      url += `?${queryString}`
    }

    return url
  } /**
     * 直接获取 Emby 图片内容
     * @param itemId 项目 ID
     * @param imageType 图片类型 (Primary, Backdrop, Logo等)
     * @param options 可选参数对象
     * @param options.tag 可选的图片标签
     * @param options.maxWidth 可选的最大宽度
     * @param options.maxHeight 可选的最大高度
     * @param options.quality 可选的图片质量
     * @returns Promise<Blob> 图片数据的 Blob 对象
     */

  async getImageContent(
    itemId: string,
    imageType: string,
    options?: {
      tag?: string
      maxWidth?: number
      maxHeight?: number
      quality?: number
    },
  ): Promise<Blob> {
    const url = this.getImageUrl(itemId, imageType, options)

    // 使用原生 fetch API 获取图片数据
    const response = await fetch(url, {
      // 凭证策略，确保发送Cookie (如果有)
      credentials: 'include',
    })

    if (!response.ok) {
      throw new Error(`获取图片失败: ${response.status} ${response.statusText}`)
    }

    return await response.blob()
  }

  /**
   * 获取 Emby 图片并创建一个对象 URL (可以直接用于 img src)
   * @param itemId 项目 ID
   * @param imageType 图片类型 (Primary, Backdrop, Logo等)
   * @param options 图片参数对象
   * @param options.tag 可选的图片标签
   * @param options.maxWidth 可选的最大宽度
   * @param options.maxHeight 可选的最大高度
   * @param options.quality 可选的图片质量
   * @returns Promise<string> 图片的对象 URL
   */
  async getImageObjectUrl(
    itemId: string,
    imageType: string,
    options?: {
      tag?: string
      maxWidth?: number
      maxHeight?: number
      quality?: number
    },
  ): Promise<string> {
    const blob = await this.getImageContent(itemId, imageType, options)
    return URL.createObjectURL(blob)
  }
}

// 导出 Emby API 实例
export const embyAPI = new EmbyAPI()
